examples/gpt/deepspeed_hf.py

```diff
@@ -125,26 +130,36 @@ def train(args):
         batch_size = 16 if batch_size is None else batch_size
         micro_batch_size = 4 if micro_batch_size is None else micro_batch_size
         ds_config_dict = get_ds_config(
-            batch_size, micro_batch_size, True, False, "Pipeline", False
+            batch_size, micro_batch_size, args.fp16, False, "Pipeline", args.bf16
         )
         loss_fct = ParallelCrossEntropy(group=group)
 
         def loss_fn(outputs, labels):
-            prediction_scores = outputs
-            shifted_prediction_scores = prediction_scores[:, :-1, :].contiguous()
-            labels = labels[:, 1:].contiguous()
+            prediction_scores = outputs.to(torch.float32)
+            shifted_prediction_scores = prediction_scores[..., :-1, :].contiguous()
+            labels = labels[..., 1:].contiguous()
             lm_loss = loss_fct(shifted_prediction_scores, labels)
             lm_loss = lm_loss.contiguous().mean()
             return lm_loss
 
-        model, _ = slapo.build(
+        (model, _), sch = slapo.build(
             sch,
             topology=topology,
             target="deepspeed",
             config=ds_config_dict,
             loss_fn=loss_fn,
             init_weights=model._init_weights,
         )
+
+        group_src_rank = _get_global_rank(sch.group, 0)
+
+        def broadcast_input(module, _inputs):
+            for inp in _inputs:
+                dist.broadcast(inp, src=group_src_rank, group=sch.group)
+            return _inputs
+
+        if model.mpu.get_pipe_parallel_rank() == 0:
+            sch["submod_0"].sync(mode="fwd_pre", sync_op_or_fn=broadcast_input)
```

Comment:
This programming model is weird. Users should not add more schedules after `.build`. In other word, `.build` should not return schedule. Since users have used this primitive in schedule, we should make sure this hook is preserved when tracing the module instead of asking users to specify again.

Reply:
Please check out #27 and see if that meets the requirements of this PR. If so, please rebase this PR after merging that.